### ISSUE TITLE
[java] ImmutableField: Do not flag fields in @Entity

### ISSUE THREAD

**Issue by numeralnathan** [pmd/pmd#3823]
**Affects PMD Version:** 6.43.0

**Rule:** [ImmutableField](https://pmd.github.io/pmd/pmd_rules_java_design.html#immutablefield)

**Description:**

JPA classes do not allow for `final` fields.  The primary key fields (i.e., annotated with `@javax.persistence.Id`) are sometimes implemented as immutable (i.e., no setters).  `ImmutableField` rule flags that these fields could be `final`.  Please fix the rule to not operate on fields annotated with `@javax.persistence.Id`.

**Code Sample demonstrating the issue:**

```
@Entity
@Table(name = "my_table")
public class MyTable implements Serializable
{
   private static final long serialVersionUID = 1L;

   @Id
   @Column(name = "directory", nullable = false)
   private long m_id;  // PMD Violation: ImmutableField

   // other fields

   /**
    * For JPA
    */
   public MyTable()
   {
      // nothing to do
   }

   public long getId()
   {
      return m_id;
   }
}
```

**Expected outcome:**

PMD reports a violation at line `private long m_id;  // PMD Violation: ImmutableField`, but that's wrong. That's a false positive.

**Running PMD through:** Eclipse PMD Plugin

**Comment by oowekyala** [comment]
Thanks for the report. Pardon my ignorance, but does a field need both `@Id` and `@Entity` on the class to fall in the case you describe? What about fields in an `@Entity` and without `@Id`, or the reverse?

**Comment by numeralnathan** [comment]
@oowekyala I apologize.  I removed `@Entity` from the original post.  It doesn't make sense.  I was looking at another ticket when I created this one.

`@Entity` goes on classes.  `@Id` goes on fields.  This request is to disable `ImmutableField` for fields annotated with `@Id` since `@Id` is generated one time and then never changed again.  However, fields cannot be final according to JPA rules.
